library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java

```diff
@@ -0,0 +1,88 @@
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.chat.api;
+
+import org.quiltmc.qsl.chat.api.types.AbstractChatMessage;
+import org.quiltmc.qsl.chat.impl.ChatEventBooleanImpl;
+import org.quiltmc.qsl.chat.impl.InternalChatEventCallbackConverters;
+import org.quiltmc.qsl.chat.impl.ChatEventImpl;
+
+/**
+ * Events for modifying, canceling, and listening for various chat messages.
+ * Events are always executed in the order {@link #MODIFY} -> {@link #CANCEL} -> {@link #BEFORE_PROCESS} -> {@link #AFTER_PROCESS}, unless a mod cancels
+ * the message, in which case {@link #CANCELLED} is invoked instead of {@link #BEFORE_PROCESS} and {@link #AFTER_PROCESS}.
+ * <p>
+ * When listening, you will only receive messages that match the provided types of your listener. If you do not provide any of a certain meta message type,
+ * then any messages for that meta type will be passed along as long as they match your other specified types.
+ */
+public final class QuiltChatEvents {
+	private QuiltChatEvents() {}
+
+	/**
+	 * An event that allows you to modify a message before further processing by returning a new one to replace it.
+	 * The usage of `withX` methods is recommended.
+	 */
+	public static final ChatEvent<Modify, AbstractChatMessage<?>> MODIFY = new ChatEventImpl<>(true, InternalChatEventCallbackConverters::modifyToHook);
+
+	/**
+	 * An event that allows you to cancel a message by returning {@code true}, or {@code false} to allow it to continue through.
+	 *
+	 * @see #CANCELLED
+	 */
+	public static final ChatEvent<Cancel, Boolean> CANCEL = new ChatEventBooleanImpl<>(InternalChatEventCallbackConverters::cancelToHook);
+
+	/**
+	 * An event that allows you to listen for messages that have been canceled.
+	 *
+	 * @see #CANCEL
+	 */
+	public static final ChatEvent<Listen, Void> CANCELLED = new ChatEventImpl<>(false, InternalChatEventCallbackConverters::listenToHook);
+
+	/**
+	 * Before (usually) vanilla does any standard processing with this message. Mods may execute other behavior before or after this event.
+	 */
+	public static final ChatEvent<Listen, Void> BEFORE_PROCESS = new ChatEventImpl<>(false, InternalChatEventCallbackConverters::listenToHook);
+
+	/**
+	 * After (usually) vanilla does any standard processing with this message. Mods may execute other behavior before or after this event.
+	 */
+	public static final ChatEvent<Listen, Void> AFTER_PROCESS = new ChatEventImpl<>(false, InternalChatEventCallbackConverters::listenToHook);
+
+	/**
+	 * A {@link FunctionalInterface} that is used with {@link #MODIFY} to modify messages.
+	 */
+	@FunctionalInterface
+	public interface Modify {
+		AbstractChatMessage<?> modifyMessage(AbstractChatMessage<?> abstractMessage);
+	}
+
+	/**
+	 * A {@link FunctionalInterface} that is used with {@link #CANCEL} to cancel messages.
+	 */
+	@FunctionalInterface
+	public interface Cancel {
+		boolean shouldCancel(AbstractChatMessage<?> abstractMessage);
```

Comment:
I'm wondering if this method is ok, it having AbstractChatMessage as an argument should reduce conflict risks but the name might be weird in the middle of a listener class.

If my message was confusing: imagine a class that implements this interface and lot of others and add an instance of it to the event, this method name might get confusing in such class.

Reply:
Yeah makes sense, ill fix those up